storage/named_query.go

```diff
@@ -83,6 +84,23 @@ var namedQueries = map[NamedQuery]string{
 	OR ((v.platform_id = :platform_id OR v.platform_id IS NULL) 
 		AND NOT EXISTS(SELECT vl.visibility_id FROM visibility_labels vl WHERE vl.visibility_id = v.id) 
 		AND sp.catalog_name = :service_plan_name AND so.catalog_name = :service_offering_name)`,
+	QueryForSharedInstances: `
+	select service_instances.*,
+	service_instance_labels.id         "service_instance_labels.id",
+	service_instance_labels.key        "service_instance_labels.key",
+	service_instance_labels.val        "service_instance_labels.val",
+	service_instance_labels.created_at "service_instance_labels.created_at",
+	service_instance_labels.updated_at "service_instance_labels.updated_at"
+	from service_instances
+	inner join service_plans on service_plans.id = service_instances.service_plan_id
+	inner join service_instance_labels on service_instances.id = service_instance_labels.service_instance_id
+	where service_instances.shared=true and
+		  service_plans.service_offering_id = :offering_id and
+		  exists (
+			SELECT * FROM service_instance_labels 
+			WHERE 	service_instance_labels.key = :tenant_identifier and 
+					service_instance_labels.val = :tenant_id
+		  )`,
```

Comment:
are you sure this query works?
you need to add in exists and service_instance_labels.service_instance_id = service_instances.id

Reply:
I think you need much more tests

Reply:
I just found it on an ownership test, fixing it.

Reply:
added tests of service offering and ownership validation for each selector:

when shared instance owned by a different tenant
- should fail to provision with plan name selector
- should fail to provision with instance name selector
- should fail to provision with instance id=(*) selector
- should fail to provision with a label selector

when a shared instance is of a different service offering
- should fail to provision with plan name selector
- should fail to provision with instance name selector
- should fail to provision with a label selector

Reply:
it won't check your fix
you need to check multiple labels selector

Reply:
we do have 
`when the tenant has many shared instances` 
-> `it creates a reference to an instance by label selector`
-> `it should fail to create a reference with labels selector if one of the labels does not match`

Reply:
also added now:
-> `it creates a reference to an instance by label selector if at least one label key matches a single selector value`
-> `it creates a reference to an instance by label selector if at least one label matches`

Reply:
yes you right the query bug is for tenancy check, the exists will return always  true

Reply:
also added now:
-> `it creates a reference to an instance by label selector if the instance has at least one selector label value`
-> `it creates a reference to an instance by label selector if the instance has at least one selector label (with its all values)`

Reply:
The description of test unclear what does it mean?
"if the instance has at least one selector label value"
and this:
"if the instance has at least one selector label (with its all values)"
Describe("creating reference")
When("selector by labels")
It(should succeed blah blah)
If I don't understand no one will understand

Reply:
"at least" is confusing when you need all labels to match

Reply:
Do you think we need tests for both OSB and SMAAP flows?
I have some duplications that I can delete, I think we can keep only the SMAAP tests since it's a common util file.

Reply:
I will rename the tests

Reply:
you need both for basic flows (byID, byLabels, byPlan ....)
and you can create dedicated test file to run all cases on the common logic to test ExtractReferenceID

Reply:
OK great, this is the current status

Reply:
we have this for the OSB flow: 
![image](https://user-images.githubusercontent.com/33447266/120764039-7ce5ea80-c520-11eb-9ce0-c2a204877942.png)

Reply:
btw, we don't have a dedicated test file for the selectors.
we run it from `service_instance_tests.go`
under the `instance sharing` section